File: src/main/java/com/hodvidar/insuratradeflow/mvc/api/controller/InsurancePolicyController.java
package com.hodvidar.insuratradeflow.mvc.api.controller;

import com.hodvidar.insuratradeflow.mvc.api.dto.InsurancePolicyDto;
import com.hodvidar.insuratradeflow.mvc.business.mapper.InsurancePolicyMapper;
import com.hodvidar.insuratradeflow.mvc.business.service.InsurancePolicyService;
import com.hodvidar.insuratradeflow.mvc.business.validation.InsurancePolicyValidationException;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/v1/insurance-policies")
public class InsurancePolicyController {

    private final InsurancePolicyService insurancePolicyService;

    private final InsurancePolicyMapper insurancePolicyMapper;

    @Autowired
    private InsurancePolicyController(final InsurancePolicyService insurancePolicyService,
                                      final InsurancePolicyMapper insurancePolicyMapper) {
        this.insurancePolicyService = insurancePolicyService;
        this.insurancePolicyMapper = insurancePolicyMapper;
    }

    @PostMapping
    public ResponseEntity<InsurancePolicyDto> createInsurancePolicy(@Valid @RequestBody InsurancePolicyDto insurancePolicy) throws InsurancePolicyValidationException {
        return ResponseEntity.status(HttpStatus.CREATED)
                .body(insurancePolicyMapper.modelToDto(insurancePolicyService.createInsurancePolicy(insurancePolicy)));
    }

    @GetMapping
    public ResponseEntity<Page<InsurancePolicyDto>> getAllInsurancePolicies(
            @RequestParam(defaultValue = "0") int page,
            @RequestParam(defaultValue = "100") int size,
            @RequestParam(defaultValue = "name,asc") String[] sort) {

        Sort sorting = Sort.by(sort[0]);
        if (sort.length > 1 && sort[1].equalsIgnoreCase("desc")) {
            sorting = sorting.descending();
        } else {
            sorting = sorting.ascending();
        }

        Pageable pageable = PageRequest.of(page, size, sorting);
        Page<InsurancePolicyDto> policyPage = insurancePolicyService.getAllInsurancePolicies(pageable)
                .map(insurancePolicyMapper::modelToDto);

        return ResponseEntity.ok(policyPage);
    }

    @GetMapping("/{id}")
    public InsurancePolicyDto getInsurancePolicy(@PathVariable Long id) {
        return insurancePolicyMapper.modelToDto(insurancePolicyService.getInsurancePolicyById(id));
    }

    @PutMapping("/{id}")
    public InsurancePolicyDto updateInsurancePolicy(@PathVariable Long id, @RequestBody InsurancePolicyDto insurancePolicyDto) throws InsurancePolicyValidationException {
        return insurancePolicyMapper.modelToDto(insurancePolicyService.updateInsurancePolicy(id, insurancePolicyDto));
    }

    @DeleteMapping("/{id}")
    public InsurancePolicyDto deleteInsurancePolicy(@PathVariable Long id) {
        return insurancePolicyMapper.modelToDto(insurancePolicyService.deleteInsurancePolicy(id));
    }
}

File: backend/src/node/api/controllers/insurance-policy.controller.ts
import {Controller, Get, Post, Put, Delete, Body, Param, Query, HttpStatus, HttpCode} from '@nestjs/common';
import * as domain from '../../domain';
import * as api from '../../api';

@Controller('api/v1/insurance-policies')
export class InsurancePolicyController {
    constructor(
        private readonly insurancePolicyService: domain.InsurancePolicyService,
        private readonly insurancePolicyMapper: domain.InsurancePolicyMapper
    ) {
    }

    @Post()
    async createInsurancePolicy(
        @Body() insurancePolicyDto: api.InsurancePolicyDto,
    ): Promise<api.InsurancePolicyDto> {
        const createdPolicy = await this.insurancePolicyService.createInsurancePolicy(insurancePolicyDto);
        return this.insurancePolicyMapper.modelToDto(createdPolicy);
    }

    @Get()
    async getAllInsurancePolicies(
        @Query() pageOptionsDto: api.PageOptionsDto,
    ): Promise<api.PageDto<api.InsurancePolicyDto>> {
        const [policies, itemCount] = await this.insurancePolicyService.getAllInsurancePolicies(pageOptionsDto);

        const pageMetaDto = new PageMetaDto({
            itemCount,
            pageOptionsDto,
        });

        return new PageDto(
            policies.map(policy => this.insurancePolicyMapper.modelToDto(policy)),
            pageMetaDto,
        );
    }

    @Get(':id')
    async getInsurancePolicy(@Param('id') id: number): Promise<api.InsurancePolicyDto | null>  {
        const policy = await this.insurancePolicyService.getInsurancePolicyById(id);
        if (!policy) {
            throw new Error(`Insurance Policy with id ${id} not found`);
        }
        return this.insurancePolicyMapper.modelToDto(policy);
    }

    @Put(':id')
    async updateInsurancePolicy(
        @Param('id') id: number,
        @Body() insurancePolicyDto: api.InsurancePolicyDto,
    ): Promise<api.InsurancePolicyDto | null | undefined> {
        const updatedPolicy = await this.insurancePolicyService.updateInsurancePolicy(id, insurancePolicyDto);
        return this.insurancePolicyMapper.modelToDto(updatedPolicy);
    }

    @Delete(':id')
    @HttpCode(HttpStatus.NO_CONTENT) // 204 for successful deletion
    async deleteInsurancePolicy(@Param('id') id: number): Promise<api.InsurancePolicyDto | null> {
        const deletedPolicy = await this.insurancePolicyService.deleteInsurancePolicy(id);
        if (!deletedPolicy) {
            throw new Error(`Insurance Policy with id ${id} not found`);
        }
        return deletedPolicy ? this.insurancePolicyMapper.modelToDto(deletedPolicy) : null;
    }
}